### ISSUE TITLE
Read/Write-Only bugs in generated code when referenced in collections

### ISSUE THREAD

**Issue by ehvattum** [oazapfts/oazapfts#446]
Hello,

It seems the readonly and writeonly support introduced in 57a4e01df03 have broken some cases of code generation.

We have a type `Product` with a readonly `id` property.
We have a type `PagedListOfProduct` with a property `items` that is a collection of `Product`s.

When using oazapfts version 4.7.3, the generated PagedListOfProduct type references the _default_ `Product` type, while I would expect it to reference the generated `ProductRead` type instead.

Schema:

```json
{
    "openapi": "3.0.0",    
    "servers": [
        {
            "url": "localhost:5001"
        }
    ],
    "paths": {
        "/api/products": {
            "get": {
                "tags": [
                    "Products"
                ],
                "operationId": "Products_GetAll",
                "parameters": [
                    {
                        "name": "PageIndex",
                        "in": "query",
                        "schema": {
                            "type": "integer",
                            "format": "int32",
                            "default": 0,
                            "maximum": 2147483647,
                            "minimum": 0
                        }
                    },
                    {
                        "name": "PageSize",
                        "in": "query",
                        "schema": {
                            "type": "integer",
                            "format": "int32",
                            "default": 15,
                            "maximum": 2147483647,
                            "minimum": 1
                        }
                    }
                ],
                "responses": {
                    "200": {
                        "description": "",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/PagedListOfProduct"
                                }
                            }
                        }
                    }
                }
            }
        }
    },
    "components": {
        "schemas": {
            "PagedListOfProduct": {
                "type": "object",
                "additionalProperties": false,
                "required": [
                    "items",
                    "totalPages",
                    "pageIndex",
                    "pageSize",
                    "totalCount"
                ],
                "properties": {
                    "items": {
                        "type": "array",
                        "items": {
                            "$ref": "#/components/schemas/Product"
                        }
                    },
                    "totalPages": {
                        "type": "integer",
                        "format": "int32"
                    },
                    "pageIndex": {
                        "type": "integer",
                        "format": "int32"
                    },
                    "pageSize": {
                        "type": "integer",
                        "format": "int32"
                    },
                    "totalCount": {
                        "type": "integer",
                        "format": "int32"
                    }
                }
            },
            "Product": {
                "type": "object",
                "additionalProperties": false,
                "required": [
                    "id",
                    "name",
                    "description",
                    "currency",
                    "producerID"
                ],
                "properties": {
                    "id": {
                        "type": "string",
                        "readOnly": true,
                        "format": "guid",
                        "minLength": 1
                    },
                    "name": {
                        "type": "string",
                        "minLength": 1
                    },
                    "description": {
                        "type": "string",
                        "minLength": 1
                    },
                    "currency": {
                        "type": "string",
                        "minLength": 1
                    },
                    "producerID": {
                        "type": "string",
                        "readOnly": true,
                        "format": "guid",
                        "minLength": 1
                    }
                }
            }
        }
    }
}
```

The generated code:

```ts
export interface Product {
    name: string;
    description: string;
    currency: string;
}
export type PagedListOfProduct = {
    items: Product[];
    totalPages: number;
    pageIndex: number;
    pageSize: number;
    totalCount: number;
};
export interface ProductRead extends Product {
    id: string;
    producerID: string;
}
```

**Comment by Xiphe** [comment]
Hi @ehvattum thank's for reporting this! 

When I understand this correctly, the misbehavior with this is that `Product` in `PagedListOfProduct` now doesn't have an `id` nor a `producterID` property, right? Or is there something else?

**Comment by ehvattum** [comment]
That is correct. I would expect those properties to be present.

**Comment by Xiphe** [comment]
Hey @eric-carlsson thanks for chiming in here 👍

Can you please check if the preview version of the changes I'm about to merge with https://github.com/oazapfts/oazapfts/pull/451 solve the problem on your end by installing `oazapfts@4.8.0-read-write-fixes.2`?

You can also opt-out of the read-/writeonly behaviour with the (currently hidden) [`--mergeReadWriteOnly` flag](https://github.com/oazapfts/oazapfts/pull/450#issuecomment-1657905684)

Please let me know when any of this helps. 

_Also I'll be on vacation for the next ~10days and will not be on GitHub during that time_

**Comment by eric-carlsson** [comment]
@Xiphe Yes, that patch seems to have resolved the issues I was experiencing! The procedures now return the `XxxRead` types, which is what I would expect.

Yes, I read that the `--mergeReadWriteOnly`option was added and I tried it out, but frankly the support for read/write-only attributes was one of the main attraction points of this library compared to other similar ones. It would of course be possible to define separate models in the OpenAPI schema for read/write, and use any of the other code gen libraries, but I preferably want to avoid that since it largely defeats the purpose of reusable schema components. :)

As a side note, some OpenAPI code gen implementations solve this issue by setting model fields that are either `readonly` or `writeOnly` as optional. The clear upside to this is that a single type can be used to represent the model in all situations, but on the other hand it's less strict in terms of type safety since it's not clear what fields are truly optional. One example of a library that does this, and that I have personally used, is [deepmap/oapi-codegen](https://github.com/deepmap/oapi-codegen) for go. I'm not saying this is better or worse, only that it's another solution to the problem.

Enjoy your vacation! :)
